### PR TITLE
Fix editing event from search room view

### PR DESCRIPTION
**Pull request by t3chguy**
Fixes https://github.com/vector-im/element-web/issues/26508

---
Here's what your changelog entry will look like:

## 🐛 Bug Fixes
 * Fix editing event from search room view ([\#11992](https://github.com/matrix-org/matrix-react-sdk/pull/11992)). Fixes vector-im/element-web#26508.

### CODE REVIEW

**Review by richvdh — CHANGES_REQUESTED** [review]
No idea what `IRoomState.search` is. Please give it a doc comment.

I don't really understand what this change is aiming to achieve. Seems like it clears the search results when you click "edit"? Which seems... odd? Could probably do with a comment, anyway.

File: src/components/structures/RoomView.tsx [hunk 1]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             case Action.EditEvent: {
                 // Quit early if we're trying to edit events in wrong rendering context
                 if (payload.timelineRenderingType !== this.state.timelineRenderingType) return;
+                if (payload.event && payload.event.getRoomId() !== this.state.roomId) {
+                    // If the event is in a different room, we need to view that room first
+                    dis.dispatch<ViewRoomPayload>({
+                        action: Action.ViewRoom,
+                        room_id: payload.even?.getRoomId(),
+                        metricsTrigger: "MessageSearch",
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1219]: why does an "edit" action cause a "MessageSearch" metricsTrigger?

File: src/components/structures/RoomView.tsx [hunk 2]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             case Action.EditEvent: {
                 // Quit early if we're trying to edit events in wrong rendering context
                 if (payload.timelineRenderingType !== this.state.timelineRenderingType) return;
+                if (payload.event && payload.event.getRoomId() !== this.state.roomId) {
+                    // If the event is in a different room, we need to view that room first
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1215]: how on earth would we get an "editevent" for an event in another room?

File: src/components/structures/RoomView.tsx [hunk 3]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
...
                 const editState = payload.event ? new EditorStateTransfer(payload.event) : undefined;
-                this.setState({ editState }, () => {
-                    if (payload.event) {
-                        this.messagePanel?.scrollToEventIfNeeded(payload.event.getId());
-                    }
-                });
+                this.setState(
+                    {
+                        editState,
+                        // Clear the search state as we don't consume editState in SearchRoomView
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1229]: what does "consuming" state mean?

**Review by richvdh — CHANGES_REQUESTED** [review]
I still don't really understand this, I'm afraid. Indeed, I have more questions than when I started.

If someone else wants to take the review and is happy with it, that's fine. Otherwise I think you're going to have to ELI5

File: src/components/structures/RoomView.tsx (hunk 1 above)
Comment (t3chguy) [reply src/components/structures/RoomView.tsx:1219]: It doesn't inherently, just in this case, will omit it for risk of overdoing it

File: src/components/structures/RoomView.tsx (hunk 2 above)
Comment (t3chguy) [reply src/components/structures/RoomView.tsx:1215]: Room > Search > All rooms - shows a timeline of events from all rooms

File: src/components/structures/RoomView.tsx [hunk 4]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
             case Action.EditEvent: {
                 // Quit early if we're trying to edit events in wrong rendering context
                 if (payload.timelineRenderingType !== this.state.timelineRenderingType) return;
+                if (payload.event && payload.event.getRoomId() !== this.state.roomId) {
+                    // If the event is in a different room (e.g. all rooms search), we need to view that room first
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1215]:
```suggestion
                    // If the event is in a different room (e.g. because the event to be edited is being displayed in the
                    // results of an all-rooms search), we need to view that room first.
```

File: src/components/structures/RoomView.tsx [hunk 5]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
...
+                    dis.dispatch<ViewRoomPayload>({
+                        action: Action.ViewRoom,
+                        room_id: payload.event?.getRoomId(),
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1218]: We already know that `payload.event` is non-nullish

```suggestion
                        room_id: payload.event!.getRoomId(),
```

File: src/components/structures/RoomView.tsx [hunk 6]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
...
                 const editState = payload.event ? new EditorStateTransfer(payload.event) : undefined;
-                this.setState({ editState }, () => {
-                    if (payload.event) {
-                        this.messagePanel?.scrollToEventIfNeeded(payload.event.getId());
-                    }
-                });
+                this.setState(
+                    {
+                        editState,
+                        // Close any ongoing search as the RoomSearchView doesn't pass editState and thus won't render the edit composer
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1229]: pass it to where?

I think I would phrase this as something like:

> If a search is active (implying that the "edit" button has been pressed on one of the events in the search result), we need to close that search, because `RoomSearchView` doesn't handle editing and won't render the composer.

File: src/components/structures/RoomView.tsx [hunk 7]
```diff
@@ -1208,12 +1211,30 @@ export class RoomView extends React.Component<IRoomProps, IRoomState> {
...
+                        search: undefined,
+                    },
+                    () => {
+                        if (payload.event) {
```
Comment (richvdh) [inline src/components/structures/RoomView.tsx:1233]: feel like maybe we should have an `if (payload.event)` at the start of the handler rather than lots of separate conditions on it, but whatever

**Review by richvdh — APPROVED** [review]
seems to make a bit more sense now
